pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/reducer/SegmentPreprocessingReducer.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.hadoop.job.reducer;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapred.AvroValue;
+import org.apache.avro.mapreduce.AvroMultipleOutputs;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.hadoop.job.JobConfigConstants.*;
+
+
+public class SegmentPreprocessingReducer<T>
+    extends Reducer<T, AvroValue<GenericRecord>, AvroKey<GenericRecord>, NullWritable> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessingReducer.class);
+
+  private AvroMultipleOutputs _multipleOutputs;
+  private AtomicInteger _counter;
+  private int _maxNumberOfRecords;
+  private String _filePrefix;
+
+  @Override
+  public void setup(Context context) {
+    LOGGER.info("Using multiple outputs strategy.");
+    Configuration configuration = context.getConfiguration();
+    _multipleOutputs = new AvroMultipleOutputs(context);
+    _counter = new AtomicInteger();
+    _maxNumberOfRecords = configuration.getInt(MAXIMUM_NUMBER_OF_RECORDS, 0);
+    LOGGER.info("Maximum number of records per file: {}", _maxNumberOfRecords);
+    _filePrefix = RandomStringUtils.randomAlphanumeric(4);
```

Comment:
why do we need this? Why is it 4?

Reply:
The purpose of having a radom prefix is to differentiate the output files generated from different reducers. The name of the output files would be like `tk2n3-r-00002`. So the output file would be listed as follow:
```
tk2n3-r-00001
tk2n3-r-00002
efh20-r-00001
efh20-r-00002
```

If we don't have this and instead have a fixed prefix, all the output files from different reducers will be mixed together, e.g. 
```
prefix1-00001
prefix1-00002
prefix1-00003
prefix2-00001
prefix2-00002
prefix2-00003
```
The reason of using 4 is that by default the name of output files like: `part-r-00001`. Just to have the similar length.

Reply:
interesting, so each reducer no longer just generates one file, how do we determine that?

Reply:
@jenniferdai If `max.num.records` isn't specified, reducer will generate only 1 output file. If so, reducer will truncate output file into multiple files.